### PR TITLE
[kube-state-metrics] - Ensure namespaces uniqueness

### PR DESCRIPTION
**Pull request by Allex1**
Ensure the namespaces list is unique

### CODE REVIEW

**Comment by Allex1** [comment]
@dotdc @mrueg @monotek @tariq1890 pls review

**Comment by dotdc** [comment]
Sure @Allex1, will look at this tonight.

File: charts/kube-state-metrics/templates/deployment.yaml [hunk 1]
```diff
@@ -74,15 +74,17 @@ spec:
...
+        {{- if .Values.namespaces }}
+        {{- range $ns := split "," .Values.namespaces }}
+        {{- $namespaces = append $namespaces (tpl $ns $) }}
+        {{- end }}
+        {{- end }}
+        {{- if .Values.releaseNamespace }}
+        {{- $namespaces = append $namespaces ( include "kube-state-metrics.namespace" . ) }}
+        {{- end }}
+        {{- if $namespaces }}
+        - --namespaces={{ $namespaces | mustUniq | join "," }}
```
Comment (dotdc) [inline charts/kube-state-metrics/templates/deployment.yaml:87]: Is it really a problem if you have the same namespace several times in this list?

File: charts/kube-state-metrics/templates/deployment.yaml [hunk 2]
```diff
@@ -74,15 +74,17 @@ spec:
...
         {{- end }}
-        {{- if and .Values.releaseNamespace .Values.namespaces}}
-        - --namespaces={{ template "kube-state-metrics.namespace" . }},{{ tpl (.Values.namespaces | join ",") $ }}
-        {{- else if .Values.releaseNamespace }}
-        - --namespaces={{ template "kube-state-metrics.namespace" . }}
-        {{- else if .Values.namespaces }}
-        - --namespaces={{ tpl (.Values.namespaces | join ",") $ }}
-        {{- end }}
-        {{- if .Values.namespacesDenylist }}
-        - --namespaces-denylist={{ tpl (.Values.namespacesDenylist | join ",") $ }}
```
Comment (dotdc) [inline charts/kube-state-metrics/templates/deployment.yaml:85]: Why are you removing this?

**Review by dotdc — COMMENTED** [review]
@Allex1, I left a few questions.

We are also missing context and an explanation of why this is needed.

File: charts/kube-state-metrics/templates/deployment.yaml (hunk 2 above)
Comment (Allex1) [reply charts/kube-state-metrics/templates/deployment.yaml]: good catch. By mistake. Added back

File: charts/kube-state-metrics/templates/deployment.yaml (hunk 1 above)
Comment (Allex1) [reply charts/kube-state-metrics/templates/deployment.yaml:87]: @dotdc from what I tested both ksm and prom-op handle duplicates just fine but I think it's good practice to handle any user error in the helm templates. Did the same for [prom-op](https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/templates/prometheus-operator/deployment.yaml#L69).

**Review by dotdc — APPROVED** [review]
LGTM

**Comment by dotdc** [comment]
Works, thank you @Allex1 for this improvement.

**Pull request merged**